include/unifex/materialize.hpp

```diff
@@ -0,0 +1,272 @@
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <unifex/async_trace.hpp>
+#include <unifex/receiver_concepts.hpp>
+#include <unifex/sender_concepts.hpp>
+#include <unifex/tag_invoke.hpp>
+#include <unifex/type_traits.hpp>
+
+#include <type_traits>
+
+namespace unifex
+{
+  namespace detail
+  {
+    template <typename Receiver>
+    class materialize_receiver {
+    public:
+      template <typename Receiver2>
+      explicit materialize_receiver(Receiver2&& receiver) noexcept(
+          std::is_nothrow_constructible_v<Receiver, Receiver2>)
+        : receiver_(static_cast<Receiver2&&>(receiver)) {}
+
+      template <
+          typename... Values,
+          std::enable_if_t<
+              std::is_invocable_v<
+                  decltype(unifex::set_value),
+                  Receiver,
+                  decltype(unifex::set_value),
+                  Values...>,
+              int> = 0>
+      void
+      set_value(Values&&... values) && noexcept(std::is_nothrow_invocable_v<
+                                                decltype(unifex::set_value),
+                                                Receiver,
+                                                decltype(unifex::set_value),
+                                                Values...>) {
+        unifex::set_value(
+            static_cast<Receiver&&>(receiver_),
+            unifex::set_value,
+            static_cast<Values&&>(values)...);
+      }
+
+      template <
+          typename Error,
+          std::enable_if_t<
+              std::is_invocable_v<
+                  decltype(unifex::set_value),
+                  Receiver,
+                  decltype(unifex::set_error),
+                  Error>,
+              int> = 0>
+      void set_error(Error&& error) && noexcept {
+        if constexpr (std::is_nothrow_invocable_v<
+                          decltype(unifex::set_value),
+                          Receiver,
+                          decltype(unifex::set_error),
+                          Error>) {
+          unifex::set_value(
+              static_cast<Receiver&&>(receiver_),
+              unifex::set_error,
+              static_cast<Error&&>(error));
+        } else {
+          try {
+            unifex::set_value(
+                static_cast<Receiver&&>(receiver_),
+                unifex::set_error,
+                static_cast<Error&&>(error));
+          } catch (...) {
+            unifex::set_error(
+                static_cast<Receiver&&>(receiver_), std::current_exception());
+          }
```

Comment:
Conditionally wrapping in `try`/`catch` should not be necessary. If everything in a `try` block is `noexcept`, the compiler is able to eliminate the overhead of the `try`/`catch`.

Reply:
This is trying to avoid instantiating the `set_error()` path if it will never be called.